src/services/refactors/annotateWithTypeFromJSDoc.ts

```diff
@@ -0,0 +1,224 @@
+        const decl = findAncestor(node, isTypedNode);
+        if (decl && !decl.type) {
+            const type = getJSDocType(decl);
+            const returnType = getJSDocReturnType(decl);
+            const annotate = (returnType || type && decl.kind === SyntaxKind.Parameter) ? annotateFunctionFromJSDoc :
+                type ? annotateTypeFromJSDoc :
+                    undefined;
+            if (annotate) {
+                return [{
+                    name: annotate.name,
+                    description: annotate.description,
+                    actions: [
+                        {
+                            description: annotate.description,
+                            name: actionName
+                        }
+                    ]
+                }];
+            }
+        }
+    }
+
+    function getEditsForAnnotation(context: RefactorContext, action: string): RefactorEditInfo | undefined {
+        if (actionName !== action) {
+            Debug.fail(`actionName !== action: ${actionName} !== ${action}`);
+            return undefined;
+        }
+
+        const sourceFile = context.file;
+        const token = getTokenAtPosition(sourceFile, context.startPosition, /*includeJsDocComment*/ false);
+        const decl = findAncestor(token, isTypedNode);
+        const jsdocType = getJSDocReturnType(decl) || getJSDocType(decl);
+        if (!decl || !jsdocType || decl.type) {
+            Debug.fail(`!decl || !jsdocType || decl.type: !${decl} || !${jsdocType} || ${decl.type}`);
+            return undefined;
+        }
+
+        const changeTracker = textChanges.ChangeTracker.fromContext(context);
+        changeTracker.replaceRange(sourceFile, { pos: decl.getStart(), end: decl.end }, addType(decl, transformJSDocType(jsdocType) as TypeNode));
+        return {
+            edits: changeTracker.getChanges(),
+            renameFilename: undefined,
+            renameLocation: undefined
+        };
+    }
+
+    function getEditsForFunctionAnnotation(context: RefactorContext, action: string): RefactorEditInfo | undefined {
+        if (actionName !== action) {
+            Debug.fail(`actionName !== action: ${actionName} !== ${action}`);
+            return undefined;
+        }
+
+        const sourceFile = context.file;
+        const token = getTokenAtPosition(sourceFile, context.startPosition, /*includeJsDocComment*/ false);
+        const decl = findAncestor(token, isFunctionLikeDeclaration);
+        const changeTracker = textChanges.ChangeTracker.fromContext(context);
+        changeTracker.replaceRange(sourceFile, { pos: decl.getStart(), end: decl.end }, addTypesToFunctionLike(decl));
+        return {
+            edits: changeTracker.getChanges(),
+            renameFilename: undefined,
+            renameLocation: undefined
+        };
+    }
+
+    function isTypedNode(node: Node): node is DeclarationWithType {

```

Comment:
TypeNode has a rather well defined meaning in other parts of the system. so i would make it `isDeclarationWithType` instead.

Reply:
Thanks, that name is much better.